src/Bedrock.Framework.Experimental/Transports/WebSockets/WebSocketPayloadEncoder.cs

```diff
@@ -0,0 +1,152 @@
+        {
+            var lengthRemaining = (long)length;
+            position = input.Start;
+
+            if(input.IsSingleSegment)
+            {
+                var bytesToRead = (int)Math.Min((long)lengthRemaining, input.First.Length);
+                MaskUnmaskSpan(input.FirstSpan, bytesToRead, useSimd);
+
+                position = new SequencePosition(input.First, bytesToRead);
+                return lengthRemaining - bytesToRead;
+            }
+
+            foreach (var memory in input)
+            {
+                var bytesToRead = (int)Math.Min((long)lengthRemaining, memory.Length);
+                MaskUnmaskSpan(memory.Span, bytesToRead, useSimd);
+
+                position = new SequencePosition(memory, bytesToRead);
+                lengthRemaining -= bytesToRead;
+            }
+
+            return lengthRemaining;
+        }
+
+        /// <summary>
+        /// Masks or unmasks a WebSocket payload span.
+        /// </summary>
+        /// <param name="span">The span to mask or unmask.</param>
+        /// <param name="bytesToRead">The number of bytes to read from the span.</param>
+        /// <param name="useSimd">Whether or not to use SIMD accelerated masking.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe void MaskUnmaskSpan(in ReadOnlySpan<byte> span, long bytesToRead, bool useSimd)
+        {
+            var maskingKey = _maskingKey;
+
+            fixed (byte* dataStartPtr = &MemoryMarshal.GetReference(span))
+            {
+                byte* dataPtr = dataStartPtr;
+                byte* dataEndPtr = dataStartPtr + bytesToRead;
+                byte* maskPtr = (byte*)&maskingKey;
+
+                //If there is at least one int of data available
+                if (dataEndPtr - dataStartPtr >= sizeof(int))
+                {
+                    //Start by aligning to an int boundary, makes aligned loads faster later
+                    while ((ulong)dataPtr % sizeof(int) != 0)
+                    {
+                        Debug.Assert(dataPtr < dataEndPtr);
+
+                        *dataPtr++ ^= maskPtr[_currentMaskIndex];
+                        _currentMaskIndex = (_currentMaskIndex + 1) & 3; //Modulus shortcut for power of 2 values for perf, same as _currentMaskIndex % 4
+                    }
+
+                    //We may have moved forward an uneven number of bytes along the masking key,
+                    //generate a new masking key that has those bytes rotated for whole-int unmasking
+                    int alignedMask = (int)BitOperations.RotateRight((uint)_maskingKey, _currentMaskIndex * 8);
+
+                    //If we found we should use SIMD and there is sufficient data to read
+                    if (useSimd && (dataEndPtr - dataPtr) >= Vector<byte>.Count)
+                    {
+                        //Align by whole ints to full SIMD load boundary to avoid a perf penalty for unaligned loads
```

Comment:
As side note:
On recent x86 hardware unaligned loads aren't penaltized as long they don't cross a page-boundary.
`Vector<T>` is also supported on non-x86, which may have a penalty for unaligned loads, so the comment is true.